Lombiq.Tests.UI/Services/AtataFactory.cs

```diff
@@ -24,62 +25,82 @@ public static class AtataFactory
             .UseUtcTimeZone();
 
+        builder.LogConsumers.AddDebug();
+        builder.LogConsumers.Add(new TestOutputLogConsumer(testOutputHelper));
+
         configuration.AtataConfiguration.ContextBuilder?.Invoke(builder);
 
         return new AtataScope(builder.Build(), baseUri);
     }
+
+    private static IWebDriver CreateDriver(
+        BrowserConfiguration browserConfiguration,
+        TimeoutConfiguration timeoutConfiguration,
+        ITestOutputHelper testOutputHelper)
+    {
+        IWebDriver CastDriverFactory<T>(Func<BrowserConfiguration, TimeSpan, T> factory)
+            where T : IWebDriver =>
+            factory(browserConfiguration, timeoutConfiguration.PageLoadTimeout);
+
+        // Driver creation can fail with "Cannot start the driver service on http://localhost:56686/" exceptions if the
+        // machine is under load. Retrying it here so not the whole test needs to be re-run.
+        const int maxTryCount = 3;
+        var currentTryIndex = 0;
+
+        // Force headless mode if we are in Linux without a working graphical environment.
+        if (!browserConfiguration.Headless && RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+        {
+            browserConfiguration.Headless = string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("DISPLAY"));
+        }
+
+        while (true)
+        {
+            try
+            {
+                return browserConfiguration.Browser switch
+                {
+                    Browser.Chrome => CastDriverFactory(WebDriverFactory.CreateChromeDriver),
+                    Browser.Edge => CastDriverFactory(WebDriverFactory.CreateEdgeDriver),
+                    Browser.Firefox => CastDriverFactory(WebDriverFactory.CreateFirefoxDriver),
+                    Browser.InternetExplorer => CastDriverFactory(WebDriverFactory.CreateInternetExplorerDriver),
+                    _ => throw new InvalidOperationException($"Unknown browser: {browserConfiguration.Browser}."),
+                };
+            }
+            catch (WebDriverException ex)
+            {
+                if (ex.Message.ContainsOrdinalIgnoreCase("Cannot start the driver service on") &&
+                    currentTryIndex < maxTryCount - 1)
+                {
+                    currentTryIndex++;
+                    var retryCount = maxTryCount - currentTryIndex;
+
+                    // Not using parameters because the exception can throw off the string format.
+                    testOutputHelper.WriteLineTimestampedAndDebug(
+                        "While creating the web driver failed with the following exception, it'll be retried " +
+                        FormattableString.Invariant($"{retryCount} more time(s). Exception: {ex}"));
+                }
+                else
+                {
+                    throw;
+                }
+            }
+        }
+    }
+
+    public static void SetupShellCliCommandFactory() =>
```

Comment:
I have a memory that private static methods are supposed to come before public static methods so that was intentional. [The dojo](https://orcharddojo.net/orchard-resources/CoreLibrary/DevelopmentGuidelines/CodeStyling) is a bit ambiguous on the topic now that I look at it. @Piedone what is your opinion?

Reply:
I've interpolated the rule from two rules that we do have in the Dojo styleguide:

1. Public instance methods come before private instance methods.
2. Static methods come after instance methods.

I think the point is that all private field and properties come at the beginning of the file, and all private methods at the end.

Reply:
Yeah, public static methods should come before private static ones. Please submit a PR for it: https://github.com/Lombiq/Orchard-Dojo-Library